src/library/scala/collection/mutable/OrderAwareHashMap.scala

```diff
@@ -0,0 +1,800 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+package mutable
+
+import scala.annotation.meta.{getter, setter}
+import scala.annotation.tailrec
+import scala.collection.generic.DefaultSerializationProxy
+import scala.collection.mutable
+
+/** This class implements mutable sets using a hashtable with red-black trees in the buckets for good
+  * worst-case performance on hash collisions. An `Ordering` is required for the element type. Equality
+  * as determined by the `Ordering` has to be consistent with `equals` and `hashCode`. Universal equality
+  * of numeric types is not supported (similar to `AnyRefMap`).
+  *
+  * @author  Stefan Zeiger
+  * @since   2.13
+  * @see [[http://docs.scala-lang.org/overviews/collections/concrete-mutable-collection-classes.html#hash-tables "Scala's Collection Library overview"]]
+  * section on `Hash Tables` for more information.
+  *
+  * @define Coll `mutable.OrderAwareHashMap`
+  * @define coll mutable order-aware hash set
+  * @define mayNotTerminateInf
+  * @define willNotTerminateInf
+  */
+final class OrderAwareHashMap[K, V](initialCapacity: Int, loadFactor: Double)(implicit ordering: Ordering[K])
```

Comment:
I think this should be named after what it's good for, like `CollisionProofMap` or `RobustMap`.  You want people who should be using it to pick it out as interesting when glancing through the collections names; I'm not sure "OrderAwareHashMap" immediately makes one think, "Oh, this is good for user-facing code where bad actors might use hash collisions in a DOS attack".

Reply:
Perhaps it also makes sense to put some info in the docs of mutable HashMap to say something like.
```
This map is open to hash collision DOS attacks. For a more robust HashMap implementation use @see [Name]
```

I imagine that most people will naturally go to HashMap first.

Reply:
I'm not happy with `OrderAwareHashMap`, either. Concrete implementations are usually named after the data structures, whereas abstract traits are named after the properties of the interface. Something like `TreeAssistedHashMap` would be better in my opinion. `OrderAwareMap` could be the name of a common interface for multiple implementations.